File: python/0121.买卖股票的最佳时机.py
class Solution:
    # 1. O(n) t:244ms(50%) O(1) m:23.2M(70%) DP
    def maxProfit(self, prices: List[int]) -> int:
        sm, ans = prices[0], 0
        for p in prices:
            ans = max(ans, p - sm)
            sm = min(p, sm)
        return ans

File: java/0121.买卖股票的最佳时机.java
class Solution {
    // 1. O(n) t:1ms(100%) O(1) m:57.6M(20%) DP
    public int maxProfit(int[] prices) {
        int sm = prices[0], ans = 0;
        for (int p : prices) {
            ans = Math.max(ans, p - sm);
            sm = Math.min(sm, p);
        }
        return ans;
    }
}